### ISSUE TITLE
accessible icons

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#9325]
We should consider changing the markup of some of the PatternFly icons, where the icon is used by itself. One very noteworthly place is on the updates link on the host overview. There's a bug icon that stands alone.

Details to make icons & emoji more accessible @:
- https://tink.uk/accessible-emoji/
- http://adrianroselli.com/2016/12/accessible-emoji-tweaked.html
- https://codepen.io/aardrian/pen/ENJdjN?editors=1100

The simple method in the first link is probably adaquate.

It isn't needed when the icon is just reinforcing the text.

**Comment by martinpitt** [comment]
@garrett: What do you mean with "stands alone"? It has text after it: "[fa-bug] Updates Available". Should these still get annotated?

At first sight, the 
[PatternFly-React Icon component](https://rawgit.com/patternfly/patternfly-react/gh-pages/patternfly-3/index.html?selectedKind=patternfly-react%2FWidgets%2FIcon&selectedStory=Icons&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) does not provide any additional annotations either. (Also, the system page doesn't use React, but some other pages do)

**Comment by garrett** [comment]
@martinpitt: I looked at the code specifically in the updates list and noticed that tables don't have scopes specificed, so I made a PR: https://github.com/cockpit-project/cockpit/pull/11503

I was also looking into icons not having text, but I do see the tooltip adds an `aria-describedby` when the icon is activated, making this tricker to address. I don't know of that's correct or when that is activated. I would've either added text to the icon or add an `aria-labelledby` or `aria-label`. But this might work too? I really don't know in this instance and it would have to be tested in screen readers to see what happens, I think.

**Comment by garrett** [comment]
Meanwhile: Cockpit does have a bunch of other bare icons that need to be checked, outside of the software updates page.

Meanwhile 2: The "Accessibility Insights" add-on for Chrome has found a bunch of issues:

![accessibility-insights-updates](https://user-images.githubusercontent.com/10246/55319732-d8fee100-5475-11e9-9fc7-436f09a576ff.png)

The biggie: The form elements don't have labels.
![a11y-popup](https://user-images.githubusercontent.com/10246/55319816-06e42580-5476-11e9-9039-559c3473f01f.png)

...but this part belongs in another PR. :wink: And we should run all of Cockpit through this tool to catch mistakes.

The important reason why I'm adding it here is: I tried finding the error and a solution for the icon issue and found:
- that it's noteworthy that the icon issue isn't caught by this a11y tool, sadly
- there are a bunch of other a11y-related issues that need to be fixed (across Cockpit; Software Updates is not really singled out... it's probably one of the better pages, really)

**Comment by garrett** [comment]
I've gone through Cockpit to specify exactly which icons are broken with regard to accessibility. It's not en exhaustive list, as I may have missed some, but it at least gives us something to target.

- [ ] **Shell**: privileged lock icon (`fa-unlock-alt`): `aria-hidden`, most likely (as the text next to it works without the icon)
- [ ] **Shell**: user icon (`pficon-user`): `aria-hidden`
- [ ] **Shell**: navigation status icons need `aria-label`
- [ ] **Shell**: top-level icons need `aria-hidden` (icons for host, dashboard, image builder, etc.).
- [ ] **Logs**: status icons (warnings & errors); probably fix this with `aria-label`s
- [ ] **Storage**: `fa-plus` icons need labels (probably `aria-label`, unless we want to visually add "add" text next to the +
- [ ] **Storage** & **Networking**: Logs may contain icons. The fix might be included in the log fix above.
- [ ] **Firewall**: Delete buttons need labelling (perhaps with `aria-label`)
- [ ] **Accounts** (detail page): + and - icons need `aria-label` (or text)
- [ ] **Services**: error icons (like `fa-exclamation-triangle`) need `aria-label`
- [ ] **Applications**: Refresh button needs `aria-label` or text.
- [ ] **Overview**, **Kernel Dump**, etc.: `fa-info-circle` needs `aria-label`.
- [ ] **Overview** status icons need `aria-label` or `aria-hidden` (depending on if they're redundant or not)
- [ ] **Overview** status icons probably need `role="alert"`; we could also bake this into the new overview page.

**Comment by marusak** [comment]
Checked the list and we fixed a few of them (either some things were dropped, were replaced by text or aria is handled properly).
I think we are just more aware of this when we work on pages, thus we improve it gradually.

I think we should do one PR where we fix the biggest offenders (as seen in the list) and then just count on the fact, that we eventually get to some forgotten pieces and do the right thing while on it.

**Comment by Ajjukota** [comment]
Hi @KKoukiou, could you please assign this issue to me?
I would love to work on it and contribute to the project

**Comment by jelly** [comment]
Hi @Ajjukota we don't assign issues, please to setup Cockpit and see if you can make a PR to resolve this issue.

**Comment by tripathics** [comment]
> I've gone through Cockpit to specify exactly which icons are broken with regard to accessibility. It's not en exhaustive list, as I may have missed some, but it at least gives us something to target.
> 
> * [x]  **Shell**: privileged lock icon (`fa-unlock-alt`): `aria-hidden`, most likely (as the text next to it works without the icon)
> * [x]  **Shell**: user icon (`pficon-user`): `aria-hidden`
> * [ ]  **Shell**: navigation status icons need `aria-label`
> * [x]  **Shell**: top-level icons need `aria-hidden` (icons for host, dashboard, image builder, etc.).
> * [ ]  **Logs**: status icons (warnings & errors); probably fix this with `aria-label`s - these are from https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/cockpit-components-logs-panel.jsx#L84
> * [ ]  **Storage**: `fa-plus` icons need labels (probably `aria-label`, unless we want to visually add "add" text next to the +
> * [ ]  **Storage** & **Networking**: Logs may contain icons. The fix might be included in the log fix above.
> * [x]  **Firewall**: Delete buttons need labelling (perhaps with `aria-label`)
> * [x]  **Accounts** (detail page): + and - icons need `aria-label` (or text)
> * [ ]  **Services**: error icons (like `fa-exclamation-triangle`) need `aria-label`
> * [x]  **Applications**: Refresh button needs `aria-label` or text.
> * [ ]  **Overview**, **Kernel Dump**, etc.: `fa-info-circle` needs `aria-label`.
> * [ ]  **Overview** status icons need `aria-label` or `aria-hidden` (depending on if they're redundant or not)
> * [ ]  **Overview** status icons probably need `role="alert"`; we could also bake this into the new overview page.

@garrett I found other places also where the aria-labels or are missing. Like under change system time in System, when adding NTP servers, the button element surrounding `fa-plus` icon don't have aria-label (can be found in [serverTime.js](https://github.com/cockpit-project/cockpit/blob/7e97d9775ca69f044c019d15bd316a4d436642f1/pkg/lib/serverTime.js#L572) file. Is it ok if I do them also along with the ones you mentioned?

**Comment by tripathics** [comment]
@garrett I think am almost done with all the sections that you've mentioned. 
This is my progress so far


* [x]  ***Shell***: privileged lock icon (`fa-unlock-alt`): `aria-hidden`, most likely (as the text next to it works without the icon)
* [x]  ***Shell***: user icon (`pficon-user`): `aria-hidden`
* [x]  **Shell**: navigation status icons need `aria-label`
  * pkg/shell/nav.jsx:184 - Use the status.type (=error/warning/info) variable itself
* [x]  ***Shell***: top-level icons need `aria-hidden` (icons for host, dashboard, image builder, etc.).
* [x]  **Logs**: status icons (warnings & errors); probably fix this with `aria-label`s - these are from https://github.com/cockpit-project/cockpit/blob/main/pkg/lib/cockpit-components-logs-panel.jsx#L84
  * pkg/lib/cockpit-components-logs-panel.jsx - Use aria-lael on icon itself as no span was present
* [X]  **Storage**: `fa-plus` icons need labels (probably `aria-label`, unless we want to visually add "add" text next to the +
  * pkg/storaged/stratis-details.jsx - aria-label on button (Not sure whether to use ariaLabel prop or aria-label. Went with the former.)
  * pkg/storaged/crypto-keyslots.jsx - same as above 
* [ ]  **Storage** & **Networking**: Logs may contain icons. The fix might be included in the log fix above.
  * May have been fixed as it is using the cockpit LogsPanel component
* [x]  ***Firewall***: Delete buttons need labelling (perhaps with `aria-label`)
* [x]  ***Accounts*** (detail page): + and - icons need `aria-label` (or text)
* [x]  **Services**: error icons (like `fa-exclamation-triangle`) need `aria-label`
  * pkg/systemd/services-list.jsx - Not needed in the ListingTable as it already ontains a label after icon (aria-hidden already present)
  * pkg/systemd/service-details.jsx - ServiceDetails contains label after icon beside "Status" (aria-hidden true already present)
* [x]  ***Applications***: Refresh button needs `aria-label` or text.
* [X]  **Overview**, **Kernel Dump**, etc.: `fa-info-circle` needs `aria-label`.
  * pkg/libs/serverTime.js - In ServerTimeConfig function: Add aria-label={"Server time " + header} where `header` is the Popover title in systemtime in Configuration card. Add role="alert"
  * pkg/libs/serverTime.js - In serverTime function: System time sync status  System time change setting, For "Automatically using Specific NTP servers", add (`fa-plus-icon`) and remove buttons, add aria-labels to those buttons
* [X]  **Overview** status icons need `aria-label` or `aria-hidden` (depending on if they're redundant or not)
  * Status icons already have an `aria-hidden`. For icons not redundant, `aria-label` has been added.
* [X]  **Overview** status icons probably need `role="alert"`; we could also bake this into the new overview page.
  * pkg/systemd/page-status.jsx - PageStatusNotifications function: Add role="alert" for update status icon except when it is a spinner then it is "presentation"
  
Italics are those you've done already, I haven't touched them

**Comment by garrett** [comment]
Thanks!

This is an extremely old bug (it's from the PF3 era), so icon names may have changed, icons listed might no longer exist, and there are most likely new icons too.

(Even though it's old, we still need icons to have ARIA labels if they are standalone. In _most_ cases, such as with alerts / errors, they shouldn't be standalone, but should come with text and be set up properly for that.)

**Comment by garrett** [comment]
Additionally, for any standalone icons (without text), they _also_ need tooltips in addition to ARIA. We need have to have text _somewhere_ to describe every icon, unless the icon is just used for enhancing existing text and the meaning is also conveyed to screen readers.

In other words: If it's visible by default, that's generally the best, but in some cases we can't do that (such as the icons on the navigation when there's an error on the page or in some tables where there isn't much space) they should use tooltips + ARIA.

**Comment by tripathics** [comment]
@garrett thank you for your feedback. I'm completely new to opensource and so I need some help understanding the project structure of cockpit repository. Like I couldn't open many pages like Kernel Dump in the web console on my laptop as it is disabled.

**Comment by garrett** [comment]
@tripathics: Oh! Which distribution are you using? I'd recommend Fedora for editing all pages, as some (like Kernel Dump) do not work on Debian/Ubuntu/Arch.

---

But, that's OK! We have a development container ready to use for other distributions. Details are at:

https://cockpit-project.org/external/source/HACKING

The gist of it is that you make sure you have the `toolbox` command installed, then you create a local container (command's on the above URL), and then you enter it when you want to do development work on Cockpit.

(You can edit files outside of the container, but be sure to be inside the container when you're building Cockpit. I personally run VS Codium from inside a customized toolbox container for better integration with stuff like eslint, but that's not really necessary.)

But, we recently reworked that document to be container-focused to enable everyone to be able to more easily work on Cockpit no matter where they are. (It _probably_ would also work on WSL2 on Windows and likely would also work on macOS once podman is set up with podman-machine and toolbox is installed... but AFAIK, nobody has done this. It's best to stick to Linux, regardless of the distribution, and follow the commands on that page.)

---

Anyway, when you have the container up and running locally, you probably want to use the VM development workflow (inside the container) so that you can build in the VM inside the container and then use the rsync method for `build.js` to update the code.

You'd want to set up the rsync method by modifying your ~/.ssh/config as mentioned in the "helpful tips" at https://cockpit-project.org/external/source/test/README and then you could run something like `./build.js -w -r c` or even `./build.js -w -r c kdump` specifically.

Since you have Cockpit working already locally (container or not), you could also skip most of this comment of mine and go directly to the preceding paragraph talking about setting up the ssh alias and using `build.js` in watch + rsync mode. (The `c` is the alias for the Cockpit VM that you'd have to set up. The page name, such as `kdump` is optional, but helps speed up building a little if you're just working on one page.)

**Comment by tripathics** [comment]
@garrett 
I was trying to start the image with
```
bots/vm-run -s cockpit.socket fedora-37
```

It started showing
```
# systemctl enable --now cockpit.socket
Created symlink /etc/systemd/system/sockets.target.wants/cockpit.socket → /usr/lib/systemd/system/cockpit.socket.

SSH ACCESS
  $ ssh -p 2201 -i bots/machine/identity root@127.0.0.2

COCKPIT
  http://127.0.0.2:9091

[ ^C to terminate ]
```

As per the Helpful tip of https://cockpit-project.org/external/source/test/README, I also added the snippets to my .ssh/config file. Now I can successfully login into the VM with ssh but how do I login to cockpit. What is the password of root?

**Comment by martinpitt** [comment]
@tripathics : root:foobar or admin:foobar (who has `sudo` rights)

**Comment by garrett** [comment]
@martinpitt: We probably want to document that (user/pass for testing VMs) in relevant locations in the docs... :wink:

**Comment by Chandan-Singh10** [comment]
I want to work on this issues please assign me this issues as soon as possible.

**Comment by martinpitt** [comment]
@Chandan-Singh10 Thanks for your interest, please go ahead! Coordinate here or in a pull request. But please make sure that you work through https://github.com/cockpit-project/cockpit/blob/main/HACKING.md first, to make sure that you can generally work on the Cockpit code. Please ask if you run into trouble.

**Comment by martinpitt** [comment]
This is too broad, too old, and more like a PatternFly issue.
